src/governance/net_governance.cpp

```diff
@@ -0,0 +1,449 @@
+    }
+
+    static Mutex cs_recently;
+    static std::map<uint256, std::map<CService, int64_t>> mapAskedRecently GUARDED_BY(cs_recently);
+    LOCK(cs_recently);
+
+    auto [vTriggerObjHashes, vOtherObjHashes] = m_gov_manager.FetchGovernanceObjectVotes(nMaxObjRequestsPerNode, nNow,
+                                                                                         mapAskedRecently);
+
+    if (vTriggerObjHashes.empty() && vOtherObjHashes.empty()) return -2;
+
+    LogPrint(BCLog::GOBJECT, "%s -- start: vTriggerObjHashes %d vOtherObjHashes %d mapAskedRecently %d\n", __func__,
+             vTriggerObjHashes.size(), vOtherObjHashes.size(), mapAskedRecently.size());
+
+    Shuffle(vTriggerObjHashes.begin(), vTriggerObjHashes.end(), FastRandomContext());
+    Shuffle(vOtherObjHashes.begin(), vOtherObjHashes.end(), FastRandomContext());
+
+    for (int i = 0; i < nMaxObjRequestsPerNode; ++i) {
+        uint256 nHashGovobj;
+
+        // ask for triggers first
+        if (!vTriggerObjHashes.empty()) {
+            nHashGovobj = vTriggerObjHashes.back();
+        } else {
+            if (vOtherObjHashes.empty()) break;
+            nHashGovobj = vOtherObjHashes.back();
+        }
+        bool fAsked = false;
+        for (const auto& pnode : vNodesCopy) {
+            // Don't try to sync any data from outbound non-relay "masternode" connections.
+            // Inbound connection this early is most likely a "masternode" connection
+            // initiated from another node, so skip it too.
+            if (!pnode->CanRelay() || (connman.IsActiveMasternode() && pnode->IsInboundConn())) continue;
+            // stop early to prevent setAskFor overflow
+            {
+                LOCK(::cs_main);
+                size_t nProjectedSize = m_peer_manager->PeerGetRequestedObjectCount(pnode->GetId()) + nProjectedVotes;
+                if (nProjectedSize > MAX_INV_SZ) continue;
+            }
+            // to early to ask the same node
+            if (mapAskedRecently[nHashGovobj].count(pnode->addr)) continue;
+
+            m_gov_manager.RequestGovernanceObject(pnode, nHashGovobj, connman, true);
+            mapAskedRecently[nHashGovobj][pnode->addr] = nNow + nTimeout;
+            fAsked = true;
+            // stop loop if max number of peers per obj was asked
+            if (mapAskedRecently[nHashGovobj].size() >= nPeersPerHashMax) break;
+        }
+        // NOTE: this should match `if` above (the one before `while`)
+        if (!vTriggerObjHashes.empty()) {
+            vTriggerObjHashes.pop_back();
+        } else {
+            vOtherObjHashes.pop_back();
+        }
+        if (!fAsked) i--;
+    }
+    LogPrint(BCLog::GOBJECT, "%s -- end: vTriggerObjHashes %d vOtherObjHashes %d mapAskedRecently %d\n", __func__,
+             vTriggerObjHashes.size(), vOtherObjHashes.size(), mapAskedRecently.size());
+
+    return int(vTriggerObjHashes.size() + vOtherObjHashes.size());
+}
+
+void NetGovernance::ProcessTick(CConnman& connman)
```

Comment:
This function does more than just governance sync, I don't think it's the right place for it.

Reply:
working on, PR converted to draft